### PR TITLE
fix: Make trove ids non-reusable

### LINKED ISSUE #324: Don’t allow to reuse troveId
Split from #282 as it will be tackled in a separate PR.

### PR DESCRIPTION
**Pull request by bingen**
Closes #324

### CODE REVIEW

File: contracts/src/BorrowerOperations.sol [hunk 1]
```diff
@@ -1313,6 +1314,13 @@ contract BorrowerOperations is LiquityBase, AddRemoveManagers, IBorrowerOperatio
...
     }
 
+    function _requireTroveDoesNotExists(ITroveManager _troveManager, uint256 _troveId) internal view {
+        ITroveManager.Status status = _troveManager.getTroveStatus(_troveId);
+        if (status != ITroveManager.Status.nonExistent) {
+            revert TroveExists();
+        }
+    }
+
     function _requireTroveIsOpen(ITroveManager _troveManager, uint256 _troveId) internal view {
```
Comment (danielattilasimon) [inline contracts/src/BorrowerOperations.sol:1324]: This appears to be unused now.

File: contracts/src/BorrowerOperations.sol [hunk 2]
```diff
@@ -1313,6 +1314,13 @@ contract BorrowerOperations is LiquityBase, AddRemoveManagers, IBorrowerOperatio
         return batchManager;
     }
 
+    function _requireTroveDoesNotExists(ITroveManager _troveManager, uint256 _troveId) internal view {
+        ITroveManager.Status status = _troveManager.getTroveStatus(_troveId);
+        if (status != ITroveManager.Status.nonExistent) {
+            revert TroveExists();
+        }
+    }
+
```
Comment (danielattilasimon) [inline contracts/src/BorrowerOperations.sol:1323]: `_requireTroveIsNotOpen()` appears to be unused now.

Comment (danielattilasimon) [reply contracts/src/BorrowerOperations.sol:1323]: Also the `TroveOpen()` error.

File: contracts/src/test/TestContracts/InvariantsTestHandler.t.sol [hunk 3]
```diff
@@ -653,7 +652,7 @@ contract InvariantsTestHandler is BaseHandler, BaseMultiCollateralTest {
                 assertGtDecimal(v.interestRate, MAX_ANNUAL_INTEREST_RATE, 18, "Shouldn't have failed as rate <= max");
             } else if (selector == BorrowerOperations.IsShutDown.selector) {
                 assertTrue(isShutdown[v.i], "Shouldn't have failed as branch hadn't been shut down");
-            } else if (selector == BorrowerOperations.TroveOpen.selector) {
+            } else if (selector == BorrowerOperations.TroveExists.selector) {
                 assertTrue(v.wasOpen, "Shouldn't have failed as Trove wasn't open");
```
Comment (bingen) [inline contracts/src/test/TestContracts/InvariantsTestHandler.t.sol:656]: Maybe we should change the error message here?

**Review by danielattilasimon — APPROVED** [review]
Looks good!

**Comment by danielattilasimon** [comment]
What about this TODO, do we want to add back the non-zero condition?
https://github.com/liquity/bold/blob/aef586b2ee251d58bbda3a8a8a73cad2c716ccd4/contracts/src/BorrowerOperations.sol#L360-L363

**Pull request merged**